server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -17,29 +17,40 @@
 import java.time.temporal.IsoFields;
 import java.time.temporal.TemporalAccessor;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
 import java.util.Objects;
-import java.util.function.BiConsumer;
 import java.util.function.UnaryOperator;
 
 class JavaDateFormatter implements DateFormatter {
+    @SuppressWarnings("unchecked")
+    private static <T extends DateTimeParser> T defaultRoundUp(T parser) {
```

Comment:
Ideally, we would pass the interface a `Map<TemporalField, Map<TemporalField, Long>>`, so each implementation would know how to apply each default, but that's not possible due to the negative condition on `WEEK_BASED_YEAR`.

So the alternatives are that each implementation has its own default, or it's all in one place (here or another class). We want to keep the defaults all in one place, so that means all the implementation-specific code is here. There's going to be, at most, two implementations, so I think its ok to have class-specific logic here (also, they're all package-private anyway). If we ever have more than 2, then we should look at creating a separate class to handle applying all the roundup defaults